File: src/grammar/generate.cpp
#include <jinja2cpp/template.h>
#include <yaml-cpp/yaml.h>

import jinja2_yaml_binding;

import std;

auto main(int argc, char * argv[]) -> int
{
    if (argc != 4) {
        std::println(std::cerr, "Usage: grammar_generator <grammar_yaml> <template> <output>");
        std::exit(EXIT_FAILURE);
    }

    YAML::Node grammar = YAML::LoadFile(argv[1]);

    std::ifstream template_file(argv[2]);
    std::stringstream buffer;
    buffer << template_file.rdbuf();
    template_file.close();

    jinja2::Template tmpl;

    std::ignore = tmpl.Load(buffer.str());
    jinja2::ValuesMap params{{"data", jinja2::Reflect(grammar)}};

    std::ofstream output(argv[3]);
    std::ignore = tmpl.Render(output, params);
    output.close();
}

File: src/grammar/generate.py
import sys

from jinja2 import Environment, FileSystemLoader, select_autoescape
import yaml

def main():
    if len(sys.argv) != 4:
        print("Usage: generate.py <grammar_yaml> <template> <output>", file=sys.stderr)
        sys.exit(1)
    
    with open(sys.argv[1], 'r') as grammar_file:
        grammar = yaml.safe_load(grammar_file)

    jinja_env = Environment(loader=FileSystemLoader('.'), autoescape=select_autoescape())
    template = jinja_env.get_template(sys.argv[2])

    with open(sys.argv[3], 'w') as output_file:
        output_file.write(template.render(data=grammar))

if __name__ == "__main__":
    main()
